File: java/scf-simple/src/test/java/org/mydotey/scf/source/stringproperty/cascaded/KeyCachedCascadedConfigurationSource.java
package org.mydotey.scf.source.stringproperty.cascaded;

import org.mydotey.collection.CascadedKeyMap;
import org.mydotey.scf.ConfigurationSourceConfig;

/**
 * @author koqizhao
 *
 * Jul 20, 2018
 * 
 * use CascadedKeyMap to cache the keys so as to prevent temp string creation and have better young gc
 */
public class KeyCachedCascadedConfigurationSource<C extends ConfigurationSourceConfig>
        extends CascadedConfigurationSource<C> {

    private CascadedKeyMap<String, String> _cascadedKeyMap;

    public KeyCachedCascadedConfigurationSource(CascadedConfigurationSourceConfig<C> config) {
        super(config);

        _cascadedKeyMap = new CascadedKeyMap<>();
    }

    @Override
    protected String getKey(String... keyParts) {
        return _cascadedKeyMap.getOrAdd(super::getKey, keyParts);
    }

}

File: dotnet/test/MyDotey.SCF.SimpleTest/Source/StringProperty/Cascaded/KeyCachedCascadedConfigurationSource.cs
using System;

using MyDotey.Collections.Generic;
using MyDotey.SCF.Facade;

namespace MyDotey.SCF.Source.StringProperty.Cascaded
{
    /**
     * @author koqizhao
     *
     * Jul 20, 2018
     * 
     * use CascadedKeyDictionary to cache the keys so as to prevent temp string creation and have better young gc
     */
    public class KeyCachedCascadedConfigurationSource<C> : CascadedConfigurationSource<C>
        where C : ConfigurationSourceConfig
    {
        private CascadedKeyDictionary<String, String> _cascadedKeyDictionary;

        public KeyCachedCascadedConfigurationSource(CascadedConfigurationSourceConfig<C> config)
            : base(config)
        {
            _cascadedKeyDictionary = new CascadedKeyDictionary<String, String>();
        }

        protected override String getKey(params String[] keyParts)
        {
            return _cascadedKeyDictionary.GetOrAdd(base.getKey, keyParts);
        }
    }
}